sparse/file.go

```diff
@@ -33,195 +31,56 @@ func SetupFileIO(direct bool) {
-	HashedInterval
-	Data []byte
-}
-
-// DataInterval FileInterval plus its data
-type DataInterval struct {
-	FileInterval
-	Data []byte
-}
-
-// FileReaderGroup starts specified number of readers
-func FileReaderGroup(count int, salt []byte, fileStream <-chan FileInterval, path string, unorderedStream chan<- HashedDataInterval) {
-	var wgroup sync.WaitGroup
-	wgroup.Add(count)
-	for i := 0; i < count; i++ {
-		go FileReader(salt, fileStream, path, &wgroup, unorderedStream)
-	}
-	go func() {
-		wgroup.Wait() // all the readers join here
-		close(unorderedStream)
-	}()
-}
-
-// FileWriterGroup starts specified number of writers
-func FileWriterGroup(count int, fileStream <-chan DataInterval, path string) *sync.WaitGroup {
-	var wgroup sync.WaitGroup
-	wgroup.Add(count)
-	for i := 0; i < count; i++ {
-		go FileWriter(fileStream, path, &wgroup)
-	}
-	return &wgroup
+	log.Debug("reading data from file is done")
+	return data[:n], nil
 }
 
-// FileReader supports concurrent file reading
-// multiple readres are allowed
-func FileReader(salt []byte, fileStream <-chan FileInterval, path string, wgroup *sync.WaitGroup, unorderedStream chan<- HashedDataInterval) {
-	// open file
-	file, err := fileOpen(path, os.O_RDONLY, 0)
+func WriteDataInterval(file *os.File, dataInterval Interval, data []byte) error {
+	log.Debug("writing data ...")
+	_, err := fileWriteAt(file, data, dataInterval.Begin)
 	if err != nil {
-		log.Fatal("Failed to open file for reading:", string(path), err)
+		errStr := fmt.Sprintf("Failed to write file using interval:%s, error: %s", dataInterval, err)
+		log.Error(errStr)
+		return fmt.Errorf(errStr)
 	}
-	defer file.Close()
-
-	for r := range fileStream {
-		switch r.Kind {
-		case SparseHole:
-			// Process hole
-			var hash, data []byte
-			unorderedStream <- HashedDataInterval{HashedInterval{r, hash}, data}
-
-		case SparseData:
-			// Read file data
-			data := make([]byte, r.Len())
-			status := true
-			n, err := fileReadAt(file, data, r.Begin)
-			if err != nil {
-				status = false
-				log.Error("File read error", status)
-			} else if int64(n) != r.Len() {
-				status = false
-				log.Error("File read underrun")
-			}
-			hasher := sha1.New()
-			hasher.Write(salt)
-			hasher.Write(data)
-			hash := hasher.Sum(nil)
-			unorderedStream <- HashedDataInterval{HashedInterval{r, hash}, data}
-		}
-	}
-	wgroup.Done() // indicate to other readers we are done
+	log.Debug("written data")
+	return nil
 }
 
-// FileWriter supports concurrent file reading
-// add this writer to wgroup before invoking
-func FileWriter(fileStream <-chan DataInterval, path string, wgroup *sync.WaitGroup) {
-	// open file
-	file, err := fileOpen(path, os.O_WRONLY, 0)
+func HashDataInterval(file *os.File, dataInterval Interval) ([]byte, error) {
+	data, err := ReadDataInterval(file, dataInterval)
 	if err != nil {
-		log.Fatal("Failed to open file for wroting:", string(path), err)
+		return nil, err
 	}
-	defer file.Close()
+	hasher := sha1.New()
```

Comment:
sha256/512 is slower than sha1 according to online data but should be more collision resistant. Chances of having collision in our usage is not big, so speed maybe more critical. Let's leave it for sha1. What do you think ? Here is the online data:

3 times cpu for each hash algorithms:

SHA1    11.985718965530396, 11.976419925689697, 11.86873197555542
SHA256  16.662450075149536, 21.551337003707886, 17.016510963439941
SHA512  18.339390993118286, 18.11187481880188,  18.085782051086426